File: frontend/src/utils/rut.js
// RUT chileno: espejo de backend/accounts/rut.py (mantener en sync).
// Canónico: sin puntos, con guion, K mayúscula (ej. 26711486-2). El backend
// es la frontera real; esto es solo UX (auto-formateo + validación previa).

const FACTORS = [2, 3, 4, 5, 6, 7]

// Dígito verificador (Módulo 11) del cuerpo numérico: '0'-'9' o 'K'.
export function computeDv(body) {
  if (!/^\d+$/.test(body)) return null
  let total = 0
  const reversed = body.split('').reverse()
  for (let i = 0; i < reversed.length; i += 1) {
    total += Number(reversed[i]) * FACTORS[i % 6]
  }
  const remainder = 11 - (total % 11)
  if (remainder === 11) return '0'
  if (remainder === 10) return 'K'
  return String(remainder)
}

// Normaliza a canónico y valida el DV; devuelve null si es vacío/malformado/DV incorrecto.
export function toCanonical(value) {
  if (value == null) return null
  const normalized = String(value).trim().toUpperCase().replace(/\./g, '').replace(/\s/g, '')
  if (!normalized) return null

  let body
  let dv
  if (normalized.includes('-')) {
    const parts = normalized.split('-')
    if (parts.length !== 2) return null
    ;[body, dv] = parts
  } else {
    body = normalized.slice(0, -1)
    dv = normalized.slice(-1)
  }

  if (!/^\d+$/.test(body)) return null
  if (!/^[0-9K]$/.test(dv)) return null
  if (computeDv(body) !== dv) return null
  return `${body}-${dv}`
}

// True si es un RUT válido (formato + Módulo 11). Nunca lanza.
export function isValidRut(value) {
  return toCanonical(value) !== null
}

// Auto-formateo en curso para un input controlado: agrupa el cuerpo con puntos
// y agrega el guion antes del DV (último carácter). No valida el DV.
export function formatRut(value) {
  const cleaned = String(value ?? '').toUpperCase().replace(/[^0-9K]/g, '')
  if (cleaned.length <= 1) return cleaned
  const body = cleaned.slice(0, -1).replace(/K/g, '')  // el cuerpo es solo dígitos
  const dv = cleaned.slice(-1)
  const grouped = body.replace(/\B(?=(\d{3})+(?!\d))/g, '.')
  return `${grouped}-${dv}`
}

File: backend/accounts/rut.py
"""RUT chileno: dígito verificador (Módulo 11) + normalización canónica.

Fuente única de verdad para validar/normalizar el RUT del usuario. El frontend
tiene un espejo en ``frontend/src/utils/rut.js`` (mantener ambos en sync).

Canónico almacenado: sin puntos, con guion, K mayúscula (ej. ``26711486-2``).
El login NO usa el RUT: sigue siendo por email.
"""

_FACTORS = (2, 3, 4, 5, 6, 7)


def compute_dv(body: str) -> str:
    """Dígito verificador (Módulo 11) del cuerpo numérico. Devuelve '0'-'9' o 'K'.

    ``body`` debe ser solo dígitos; ValueError si no.
    """
    if not body or not body.isdigit():
        raise ValueError('El cuerpo del RUT debe ser numérico.')
    total = sum(int(digit) * _FACTORS[i % 6] for i, digit in enumerate(reversed(body)))
    remainder = 11 - (total % 11)
    if remainder == 11:
        return '0'
    if remainder == 10:
        return 'K'
    return str(remainder)


def clean_rut(value) -> str:
    """Normaliza a canónico (``NNNNNNNN-D``) y valida el dígito verificador.

    Quita puntos y espacios, pasa K a mayúscula y agrega el guion si falta.
    Lanza ``ValueError`` si el RUT es vacío, malformado o el DV no cuadra.
    """
    if value is None:
        raise ValueError('RUT vacío.')

    normalized = str(value).strip().upper().replace('.', '').replace(' ', '')
    if not normalized:
        raise ValueError('RUT vacío.')

    if '-' in normalized:
        parts = normalized.split('-')
        if len(parts) != 2:
            raise ValueError('RUT malformado.')
        body, dv = parts
    else:
        body, dv = normalized[:-1], normalized[-1]

    if not body or not body.isdigit():
        raise ValueError('El cuerpo del RUT debe ser numérico.')
    if dv not in '0123456789K':
        raise ValueError('Dígito verificador inválido.')

    if compute_dv(body) != dv:
        raise ValueError('El dígito verificador no corresponde al RUT.')

    return f'{body}-{dv}'


def is_valid_rut(value) -> bool:
    """True si ``value`` es un RUT válido (formato + Módulo 11). Nunca lanza."""
    try:
        clean_rut(value)
    except (ValueError, TypeError):
        return False
    return True
